src/main/groovy/org/moditect/gradleplugin/ModitectExtension.groovy

```diff
@@ -0,0 +1,141 @@
+        this.dependencyResolver = new DependencyResolver(project)
+    }
+
+    void init(
+            AddMainModuleInfoTask addMainModuleInfoTask,
+            AddDependenciesModuleInfoTask addDependenciesModuleInfoTask,
+            GenerateModuleInfoTask generateModuleInfoTask,
+            CreateRuntimeImageTask createRuntimeImageTask) {
+        this.addMainModuleInfoTask = addMainModuleInfoTask
+        this.addDependenciesModuleInfoTask = addDependenciesModuleInfoTask
+        this.generateModuleInfoTask = generateModuleInfoTask
+        this.createRuntimeImageTask = createRuntimeImageTask
+
+        Util.createDirectory(workingDirectory)
+        def workingDirAsFile = workingDirectory.get().asFile
+        this.addMainModuleInfoTask.workingDirectory.set(workingDirAsFile)
+        this.addDependenciesModuleInfoTask.workingDirectory.set(workingDirAsFile)
+        this.generateModuleInfoTask.workingDirectory.set(workingDirAsFile)
+    }
+
+    private ArtifactsInfo retrieveArtifactsInfo() {
+        addDependenciesModuleInfoTask.modules.get().moduleConfigurations.each { cfg ->
+            cfg.updateFullConfiguration()
+        }
+        generateModuleInfoTask.modules.get().moduleConfigurations.each { cfg ->
+            cfg.updateFullConfiguration()
+        }
+        def cfg = project.configurations.getByName(ModitectPlugin.FULL_CONFIGURATION_NAME)
+        cfg.extendsFrom(project.configurations.getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME))
+        cfg.canBeResolved = true
+
+        Map<ModuleId, String> assignedNamesByModules = [:]
+        assignedNamesByModules.putAll(generateModuleInfoTask.assignedNamesByModules)
+        assignedNamesByModules.putAll(addDependenciesModuleInfoTask.assignedNamesByModules)
+
+        Set<ArtifactDescriptor> descriptors = []
+        cfg.resolvedConfiguration.resolvedArtifacts.each { artifact ->
+            String autoModuleName = DependencyDescriptor.getAutoModuleNameFromInputJar(artifact.file.toPath(), null);
+            def info = artifact.moduleVersion.id
+            File inputJar
+            if (autoModuleName != null) {
+                inputJar = artifact.file
+            } else {
+                String moduleName = assignedNamesByModules.get(new ModuleId(info.group, info.name))
+                inputJar = GenerateModuleInfo.createCopyWithAutoModuleNameManifestHeader(
+                        workingDirectory.get().asFile.toPath(),
+                        artifact.file.toPath(),
+                        moduleName
+                ).toFile()
+            }
+            descriptors.add(new ArtifactDescriptor(info.group, info.name, inputJar))
+        }
+        LOGGER.info "assignedNamesByModules: ${assignedNamesByModules.collect {'\n\t' + it}}"
+        LOGGER.info "artifactDescriptors: ${descriptors.collect {'\n\t' + it}}"
+        new ArtifactsInfo(assignedNamesByModules, descriptors)
+    }
+
+
+    void addMainModuleInfoData(Closure closure) {
```

Comment:
Use `Action` instead of `Closure`. This makes the extension friendly to Kotlin DSL.